atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,223 @@
+        """
+        url_components = [base_url, 'src', srpm_filename]
+        if sign_key:
+            url_components[1:1] = ['data', 'signed', sign_key]
+        return '/'.join(url_components)
+
+    def get_srpm_urls(self, build_id, sigkeys=None, insecure=False):
+        """Fetch SRPM download URLs for each image generated by a build
+
+        Build each possible SRPM URL and check if the URL is available,
+        respecting the signing intent preference order.
+
+        :param build_id: int, Koji build id of the container image we want SRPMs for
+        :param sigkeys: list, strings for keys which signed the srpms to be fetched
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        if not sigkeys:
+            sigkeys = ['']
+
+        srpm_build_paths = {}
+        archives = self.session.listArchives(build_id, type='image')
+        for archive in archives:
+            rpms = self.session.listRPMs(imageID=archive['id'])
+            for rpm in rpms:
+                rpm_hdr = self.session.getRPMHeaders(rpm['id'], headers=['SOURCERPM'])
+                srpm_filename = rpm_hdr['SOURCERPM']
+                if srpm_filename in srpm_build_paths:
+                    continue
+                rpm_build = self.session.getBuild(rpm['build_id'], strict=True)
+                path_info = get_koji_path_info(self.workflow, self.koji_fallback)
+                base_url = path_info.build(rpm_build)
+                srpm_build_paths[srpm_filename] = base_url
+
+        srpm_urls = []
+        missing_srpms = []
+        for srpm_filename, base_url in srpm_build_paths.items():
+            for sigkey in sigkeys:
+                # koji uses lowercase for paths. We make sure the sigkey is in lower case
+                url_candidate = self.assemble_srpm_url(base_url, srpm_filename, sigkey.lower())
+                req_session = get_retrying_requests_session()
+                request = req_session.head(url_candidate, verify=not insecure)
+                if request.ok:
+                    srpm_urls.append(url_candidate)
+                    break
+                self.log.debug('%s not found for signing key "%s" at %s (returned %s)',
+                               srpm_filename, sigkey, url_candidate, request.status_code)
+
+            else:
+                missing_srpms.append(srpm_filename)
+
+        if missing_srpms:
+            raise RuntimeError('Could not find files signed by any of {} for these SRPMS: {}'
+                               .format(sigkeys, srpm_filename))
+
+        return srpm_urls
+
+    def get_signing_intent(self, build_id):
+        """Get the signing intent to be used to fetch files from Koji
+
+        :param build_id: int, Koji build id of the container image which owns the signing intent
+        :return: dict, signing intent object as per atomic_reactor/schemas/config.json
+        """
+        odcs_config = get_config(self.workflow).get_odcs_config()
```

Comment:
use get_odcs_config directly like we do in plugins?

Reply:
I am doing it like this to get an ODCSConfig object, like in the resolve_composes plugin. If you prefer, I could get the odcs config and import the ODCSConfig class as well, then create the object passing the info from the config. WDYT?

Reply:
yeah I forgot that get_odcs_config isn't the public method